### PR TITLE
#163146374 Setup database models

### PR DESCRIPTION
**Pull request by Bobsar0**
#### What does this PR do?
Setup database models, validations and associations.

#### Description of Task to be Completed?
* Create all database models
* Setup model validations
* Setup model validations
* Setup model associations
* Seed sample data into database for testing
* Enable SSL on production config

#### How should this be manually tested?
* Clone the ch-setup-database-models-163146374 branch
* Install postgres on your local machine
* Create postgres user neon and grant access for creating databases.
* Login with user neon and
* Create haven_dev as a local development database
* Create haven_test as a local test database
* Run ```npm install``` command to install pg, sequelize and sequelize-cli packages
* Run ```npm run db:migrate``` and `npm run db:seed`
* Try playing around with the files in the seeders folder - the validations will set a constraint on some of the fields

#### What are the relevant pivotal tracker stories?
[Setup database models, validations and associations](https://www.pivotaltracker.com/story/show/163146374)

### CODE REVIEW

**Review by rafmme — COMMENTED** [review]
LGTM

**Review by jesseinit — COMMENTED** [review]
Looks Good to Me

File: server/migrations/20190111125359-create-role.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable no-unused-vars */
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.createTable('Roles', {
+    id: {
+      allowNull: false,
+      autoIncrement: true,
+      primaryKey: true,
+      type: Sequelize.INTEGER
+    },
+    type: {
```
Comment (seunkoko) [inline server/migrations/20190111125359-create-role.js:9]: Please map the role_id on the user's table to the `id` on the role's table. Also, make sure the type is unique.

File: server/migrations/20190111125231-create-user.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      type: Sequelize.STRING
+    },
+    img: {
+      type: Sequelize.STRING
+    },
+    notifysettings: {
+      type: Sequelize.BOOLEAN,
+      defaultValue: true
+    },
+    role: {
```
Comment (seunkoko) [inline server/migrations/20190111125231-create-user.js]: Please change to `roleId`

File: server/migrations/20190111125231-create-user.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      type: Sequelize.STRING,
+    },
+    authtype: {
```
Comment (seunkoko) [inline server/migrations/20190111125231-create-user.js]: Do unto `authtype` what you have done unto `role`

File: server/migrations/20190111125457-create-notificationtype.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable no-unused-vars */
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.createTable('Notificationtypes', {
+    id: {
+      allowNull: false,
+      autoIncrement: true,
+      primaryKey: true,
+      type: Sequelize.INTEGER
+    },
+    type: {
```
Comment (seunkoko) [inline server/migrations/20190111125457-create-notificationtype.js]: Make this unique

Comment (seunkoko) [reply server/migrations/20190111125457-create-notificationtype.js]: Also make sure it's not null

File: server/migrations/20190111130213-create-article.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
+/* eslint-disable no-unused-vars */
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.createTable('Articles', {
+    id: {
+      allowNull: false,
+      autoIncrement: true,
+      primaryKey: true,
+      type: Sequelize.INTEGER
+    },
+    slug: {
```
Comment (seunkoko) [inline server/migrations/20190111130213-create-article.js:9]: Set to not_null and unique

File: server/migrations/20190111131022-create-rating.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+/* eslint-disable no-unused-vars */
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.createTable('Ratings', {
+    id: {
+      allowNull: false,
+      autoIncrement: true,
+      primaryKey: true,
+      type: Sequelize.INTEGER
+    },
+    articleid: {
```
Comment (seunkoko) [inline server/migrations/20190111131022-create-rating.js]: Please ensure naming conventions are in camelCases... `articleId

**Review by seunkoko — CHANGES_REQUESTED** [review]

File: server/migrations/20190111131022-create-rating.js (hunk 6 above)
Comment (Bobsar0) [reply server/migrations/20190111131022-create-rating.js]: Done thanks.

File: server/migrations/20190111130213-create-article.js (hunk 5 above)
Comment (Bobsar0) [reply server/migrations/20190111130213-create-article.js:9]: Done thanks.

File: server/migrations/20190111125457-create-notificationtype.js (hunk 4 above)
Comment (Bobsar0) [reply server/migrations/20190111125457-create-notificationtype.js]: Done thanks.

File: server/migrations/20190111125231-create-user.js (hunk 3 above)
Comment (Bobsar0) [reply server/migrations/20190111125231-create-user.js]: Golden rule achieved thanks.

File: server/migrations/20190111130634-create-reply.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    id: {
+      allowNull: false,
+      autoIncrement: true,
+      primaryKey: true,
+      type: Sequelize.INTEGER
+    },
+    userId: {
+      type: Sequelize.INTEGER,
+      allowNull: false
+    },
```
Comment (Makwe-O) [inline server/migrations/20190111130634-create-reply.js:12]: I noticed that the Comment model has a column for content but the Reply model doesn't have

File: server/migrations/20190111125231-create-user.js [hunk 8]
```diff
@@ -1,46 +1,50 @@
...
-    notifysettings: {
+    notifySettings: {
       type: Sequelize.BOOLEAN,
       defaultValue: true
     },
-    role: {
-      type: Sequelize.STRING,
+    roleId: {
+      type: Sequelize.INTEGER,
+      allowNull: false
```
Comment (ebzeal) [inline server/migrations/20190111125231-create-user.js]: roleId should have a default value

File: server/migrations/20190111130634-create-reply.js (hunk 7 above)
Comment (Bobsar0) [reply server/migrations/20190111130634-create-reply.js:12]: It's been added now thanks.

File: server/migrations/20190111125231-create-user.js (hunk 8 above)
Comment (Bobsar0) [reply server/migrations/20190111125231-create-user.js]: Yes, it should. It's been added thanks.

**Comment by Makwe-O** [comment]
Looks Good To Me

**Review by Sammuel09 — COMMENTED** [review]
Looks good to me.

File: server/models/tag.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+      type: DataTypes.STRING,
+      allowNull: false,
+      unique: true
+    },
+  }, {});
+  Tag.associate = (models) => {
+    Tag.belongsToMany(models.Article, {
+      through: 'ArticleTag',
+      as: 'articles',
+      foreignKey: 'tagid'
```
Comment (rafmme) [inline server/models/tag.js]: Shouldn't this be in camelCase too?

Comment (Bobsar0) [reply server/models/tag.js]: Thanks for noticing

File: .eslintignore [hunk 10]
```diff
@@ -0,0 +1,2 @@
+server/seeders
+server/migrations
```
Comment (rafmme) [inline .eslintignore]: No new line here

File: package.json [hunk 11]
```diff
@@ -55,8 +58,7 @@
     "nyc": "^13.1.0",
     "sequelize-cli": "^5.4.0",
     "nodemon": "^1.18.9",
-    "nyc": "^13.1.0",
     "rimraf": "^2.6.3",
     "yamljs": "^0.3.0"
   }
-}
```
Comment (rafmme) [inline package.json]: New line missing here also

**Review by seunkoko — APPROVED** [review]

File: server/migrations/20190111130729-create-articleLikesDislike.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
+export default {
```
Comment (rafmme) [inline server/migrations/20190111130729-create-articleLikesDislike.js:1]: why do same files appear twice and also could this be the reason why the number of files changes has increased.
Travis build for the branch is also failing

Comment (Bobsar0) [reply server/migrations/20190111130729-create-articleLikesDislike.js:1]: You're right. It's due to Github being case-insensitive to file renaming. It's been sorted now thanks.

File: server/migrations/20190111130817-create-commentlike.js [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+      defaultValue: Sequelize.literal('CURRENT_TIMESTAMP')
+    },
+    updatedAt: {
+      allowNull: false,
+      type: Sequelize.DATE,
+      defaultValue: Sequelize.literal('CURRENT_TIMESTAMP')
+    }
+  }),
+  down: (queryInterface, Sequelize) => queryInterface.dropTable('CommentLikes')
+};
```
Comment (Sammuel09) [inline server/migrations/20190111130817-create-commentlike.js]: Hello Steve, it seems to me that this file is repeated as the same as above.

Comment (Bobsar0) [reply server/migrations/20190111130817-create-commentlike.js]: Yeah thanks for noticing Samuel. It's resolved now.

**Pull request merged**